cpp/src/io/statistics/byte_array_view.cuh

```diff
@@ -0,0 +1,187 @@
+   *
+   * The behavior is undefined if idx is out of range (i.e., if it is greater than or equal to
+   * size()).
+   *
+   * @param idx The index of the element to access.
+   * @return A reference to the idx-th element of the byte_array_view, i.e., `_data.data()[idx]`.
+   */
+  [[nodiscard]] constexpr element_type& operator[](std::size_t idx) const { return _data[idx]; }
+
+  /**
+   * @brief Returns a pointer to the beginning of the byte_array_view.
+   *
+   * @return A pointer to the first element of the byte_array_view.
+   */
+  [[nodiscard]] constexpr element_type* data() const noexcept { return _data.data(); }
+
+  /**
+   * @brief Returns the number of elements in the byte_array_view.
+   *
+   * @return The number of elements in the byte_array_view.
+   */
+  [[nodiscard]] constexpr std::size_t size() const noexcept { return _data.size(); }
+
+  /**
+   * @brief Returns the size of the byte_array_view in bytes.
+   *
+   * @return The size of the byte_array_view in bytes
+   */
+  [[nodiscard]] constexpr std::size_t size_bytes() const noexcept { return _data.size_bytes(); }
+
+  /**
+   * @brief Comparing target byte_array_view with this byte_array_view. Each byte in the array is
+   * compared.
+   *
+   * @param byte_array_view Target byte_array_view to compare with this byte_array_view.
+   * @return 0  If they compare equal.
+   *         <0 Either the value of the first byte of this byte_array_view that does not match is
+   * lower in the arg byte_array_view, or all compared bytes match but the arg byte_array_view is
+   * shorter. >0 Either the value of the first byte of this byte_array_view that does not match is
+   * greater in the arg byte_array_view, or all compared bytes match but the arg byte_array_view is
+   * longer.
+   */
+  [[nodiscard]] __device__ inline int32_t compare(byte_array_view const& rhs) const
+  {
+    auto const len1  = size_bytes();
+    auto const len2  = rhs.size_bytes();
+    auto const* ptr1 = this->data();
+    auto const* ptr2 = rhs.data();
+    if ((ptr1 == ptr2) && (len1 == len2)) { return 0; }
+    // if I am max, I am greater than the argument
+    if (ptr1 == nullptr && len1 == std::numeric_limits<std::size_t>::max()) { return 1; }
+    // if the argument is max, it is greater than me
+    if (ptr2 == nullptr && len2 == std::numeric_limits<std::size_t>::max()) { return -1; }
+    std::size_t idx = 0;
+    for (; (idx < len1) && (idx < len2); ++idx) {
+      if (ptr1[idx] != ptr2[idx]) {
+        return static_cast<int32_t>(ptr1[idx]) - static_cast<int32_t>(ptr2[idx]);
+      }
+    }
+    // if the argument ran out of data, it is less than me
+    if (idx < len1) return 1;
+    // if I ran out of data first, I am less than the argument
+    if (idx < len2) return -1;
+    return 0;
+  }
+
+  /**
+   * @brief Returns true if this string is ordered before rhs.
+   *
+   * @param rhs Target string to compare with this string.
```

Comment:
I think all mentions of string should be byte array?

Reply:
yeah, I wasn't paying attention to the comments very well here. Thanks.